src/Aspire.Hosting/Health/ResourceHealthCheckService.cs

```diff
@@ -0,0 +1,119 @@
+            //       dynamically add health checks at runtime. If this changes then we
+            //       would need to revisit this and scan for transitive health checks
+            //       on a periodic basis (you wouldn't want to do it on every pass.
+            logger.LogTrace("Stopping monitoring resource '{Resource}' has not health check annotations.", resource.Name);
+            return;
+        }
+
+        var registrationKeysToCheck = annotations.DistinctBy(a => a.Key).Select(a => a.Key).ToImmutableHashSet();
+
+        logger.LogTrace("Resource '{Resource}' transitively has the following health checks: {Keys}",
+            resource.Name,
+            registrationKeysToCheck
+            );
+
+        using var timer = new PeriodicTimer(TimeSpan.FromSeconds(5));
+
+        while (await timer.WaitForNextTickAsync(cancellationToken).ConfigureAwait(false))
+        {
+            try
+            {
+                logger.LogTrace("Checking health of resource '{Resource}'.", resource.Name);
+
+                var report = await healthCheckService.CheckHealthAsync(
+                    r => registrationKeysToCheck.Contains(r.Name),
+                    cancellationToken
+                    ).ConfigureAwait(false);
+
+                foreach (var entry in report.Entries)
+                {
+                    logger.LogTrace(
+                        "Resource '{Resource}' health check '{Check}' returned '{Status}', message was: {Message}",
+                        resource.Name,
+                        entry.Key,
+                        entry.Value.Status,
+                        entry.Value.Description
+                        );
+                }
+
+                if (_latestEvents[resource.Name].Snapshot.HealthStatus == report.Status)
+                {
+                    // If the last health status is the same as this health status then we don't need
+                    // to publish anything as it just creates noise.
+                    logger.LogTrace("Resource '{Resource}' health status has not changed from '{Status}'.", resource.Name, report.Status);
+                    continue;
+                }
+
+                await resourceNotificationService.PublishUpdateAsync(resource, s => s with
+                {
+                    HealthStatus = report.Status
+                }).ConfigureAwait(false);
+
+                if (resource.TryGetLastAnnotation<ReplicaInstancesAnnotation>(out var replicaAnnotation))
+                {
+                    foreach (var (id, _) in replicaAnnotation.Instances)
+                    {
+                        await resourceNotificationService.PublishUpdateAsync(resource, id, s => s with
+                        {
+                            HealthStatus = report.Status
+                        }).ConfigureAwait(false);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                _ = ex;
```

Comment:
?

Reply:
Cleaned this up a bit. When debugging its possible for a cancellation to occur but we don't want to stop the loop. So I watch for an operation cancelled exception but ignore it if it isn't from the stopping token for the background service.